httpx/cert_reload.go

```diff
@@ -0,0 +1,159 @@
+package httpx
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"path"
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	"github.com/ory/x/logrusx"
+)
+
+type (
+	CertLocation struct {
+		CertPath string
+		KeyPath  string
+	}
+
+	CertLoadFunc func() ([]tls.Certificate, error)
+
+	CertProvider struct {
+		certs   []tls.Certificate
+		mu      sync.Mutex
+		watcher *fsnotify.Watcher
+
+		certLoad     CertLoadFunc
+		certLocation CertLocation
+		logger       *logrusx.Logger
+	}
+)
+
+//NewCertProvider creates a tls.Certificate provider that reloads certificates on fs change
+//certs must not be empty
+func NewCertProvider(certs []tls.Certificate, certLocation CertLocation, certLoad CertLoadFunc, l *logrusx.Logger) *CertProvider {
+	ret := &CertProvider{
+		certLoad:     certLoad,
+		certLocation: certLocation,
+		logger:       l,
+	}
+	ret.load(certs)
+	ret.watchCertificatesChanges()
+
+	runtime.SetFinalizer(ret, func(ret *CertProvider) { ret.Stop() })
```

Comment:
Neat!

Reply:
In the same way in the comment above. We should implement a procedure in Hydra/Kratos to stop the watcher when shutting down the app if we don't want "runtime.SetFinalizer".